docs/developer/cluster-lifecycle-api-provider-integration.md

```diff
@@ -0,0 +1,60 @@
+
+`ProviderClusterReconciler` has to handle all the actions needed to reconcile the workload cluster to the state defined in the eks-a CRs. It has complete control over how this is implemented and there are no restrictions imposed in this regard. That said, there are 3 main tasks generic tasks that every `ProviderClusterReconciler` has to perform, in this order:
+1. Reconcile the workload cluster Control Plane to the eks-a CR specifications
+1. Reconcile the workload cluster worker nodes to the eks-a CR specifications
+1. Reconcile the CNI to the eks-a CR specifications
+
+Each provider can add any other additional steps, in order: before, after ir even in between of these 3.
+
+The CNI reconciliation should be executed by using the Cilium reconciler (and other in the future), orchestrated by the main [CNI Reconciler](https://github.com/aws/eks-anywhere/blob/main/pkg/networking/reconciler/reconciler.go). These are designed to work in a reconciliation loop provided a stable Control Plane and and a client to interact with.
+
+## Requirements
+
+* Implement the [`ProviderClusterReconciler` interface](https://github.com/aws/eks-anywhere/blob/main/pkg/controller/clusters/registry.go#L12)
+* Build and register the new reconciler in the [controller factory](https://github.com/aws/eks-anywhere/blob/main/controllers/factory.go#L170)
+* Add the necessary watches to the [cluster controller](https://github.com/aws/eks-anywhere/blob/main/controllers/cluster_controller.go#L54)
+* Update [cluster controller RBAC](https://github.com/aws/eks-anywhere/blob/main/controllers/cluster_controller.go#L94) accordingly. You might need permissions to watch the provider eks-a CRDs or read/create/update/delete the provider infra CAPI CRDs.
+* Add your provider specific API structs to [`cluster.Config`](https://github.com/aws/eks-anywhere/blob/main/pkg/cluster/config.go#L10). Add them as [child objects](https://github.com/aws/eks-anywhere/blob/main/pkg/cluster/config.go#L88) if necessary.
+* Implement the proper [client processors](https://github.com/aws/eks-anywhere/blob/main/pkg/cluster/client_builder.go#L18) for your API structs and register them in the [default `ConfigClientBuilder`](https://github.com/aws/eks-anywhere/blob/main/pkg/cluster/config.go#L10)
+* Add validation webhooks for all provider specific CRDs
+* Add mutation webhooks for all provider specific CRDs when defaults are needed
+* If context aware defaults or validations (those that require any kind of API or disk call) are needed, they should be implemented in a controller for that CRD ([example](https://github.com/aws/eks-anywhere/blob/main/controllers/snow_machineconfig_controller.go)). For any extra validations, the `ProviderClusterReconciler` will need to check if the spec of the eks-a provider CRs is valid before starting the reconciliation loop. In case they are not, bubble that error to the `Cluster.status`.
```

Comment:
I'm confused because the ProviderClusterReconciler for snow doesn't do anything. And the vsphere validations all seem to happen inside this ProviderClusterReconciler. Which approach is correct between these?

Reply:
It would also help to add a note that these ProviderClusterReconcilers need to be registered in https://github.com/aws/eks-anywhere/blob/main/manager/main.go#L112

Reply:
re preflight checks: that depends on what preflight checks we are talking about. In general, anything that validates the data stored in the objects and can prevent some kind of failure should be included.

re snow: the snow cluster reconciler checks that the spec for the machine configs is valid. In general, I wouldn't use the vsphere implementation as a guide, it needs to change in some places, snow is probably a better example. That said, each provider can have different requirements/constrains.

re registering in main: where do you see that? I probably missed it. In theory, `ProviderClusterReconcilers` only need to be registered in the factory. It's only the controller-runtime reconcilers (which in fact are controllers) that need to be registered in main (with the controller manager).